modin/backends/cudf/parser.py

```diff
@@ -0,0 +1,125 @@
+        df: The DataFrame after it has been read.
+
+    Returns:
+        A list of pandas DataFrames.
+    """
+    splits = split_result_of_axis_func_pandas(axis, num_splits, df)
+    if not isinstance(splits, list):
+        splits = [splits]
+    return splits
+
+
+def find_common_type_cat(types):
+    if all(isinstance(t, pandas.CategoricalDtype) for t in types):
+        if all(t.ordered for t in types):
+            return pandas.CategoricalDtype(
+                np.sort(np.unique([c for t in types for c in t.categories])[0]),
+                ordered=True,
+            )
+        return union_categoricals(
+            [pandas.Categorical([], dtype=t) for t in types],
+            sort_categories=all(t.ordered for t in types),
+        ).dtype
+    else:
+        return find_common_type(types)
+
+class cuDFParser(object):
+    @classmethod
+    def get_dtypes(cls, dtypes_ids):
+        return (
+            pandas.concat(cls.materialize(dtypes_ids), axis=1)
+            .apply(lambda row: find_common_type_cat(row.values), axis=1)
+            .squeeze(axis=0)
+        )
+
+    @classmethod
+    def single_worker_read(cls, fname, **kwargs):
+        ErrorMessage.default_to_pandas("Parameters provided")
+        # Use default args for everything
+        pandas_frame = cls.parse(fname, **kwargs)
+        if isinstance(pandas_frame, pandas.io.parsers.TextFileReader):
+            pd_read = pandas_frame.read
+            pandas_frame.read = lambda *args, **kwargs: cls.query_compiler_cls.from_pandas(
+                pd_read(*args, **kwargs), cls.frame_cls
+            )
+            return pandas_frame
+        elif isinstance(pandas_frame, (OrderedDict, dict)):
+            return {
+                i: cls.query_compiler_cls.from_pandas(frame, cls.frame_cls)
+                for i, frame in pandas_frame.items()
+            }
+        return cls.query_compiler_cls.from_pandas(pandas_frame, cls.frame_cls)
+
+    infer_compression = infer_compression
+
+
+class cuDFCSVParser(cuDFParser):
+    @classmethod
+    def parse(cls, fname, **kwargs):
+        warnings.filterwarnings("ignore")
+        num_splits = kwargs.pop("num_splits", None)
+        start = kwargs.pop("start", None)
+        end = kwargs.pop("end", None)
+        index_col = kwargs.get("index_col", None)
+        assigned_gpus = kwargs.pop("assigned_gpus", None)
+
+        if start is not None and end is not None:
+            put_func = cls.frame_partition_cls.put
+
+            # pop "compression" from kwargs because bio is uncompressed
+            bio = FileReader.file_open(fname, "rb", kwargs.pop("compression", "infer"))
+            if kwargs.get("encoding", None) is not None:
+                header = b"" + bio.readline()
+            else:
+                header = b""
+            bio.seek(start)
+            to_read = header + bio.read(end - start)
+            bio.close()
+            pandas_df = pandas.read_csv(BytesIO(to_read), **kwargs)
```

Comment:
What is the reason we do not use cudf's `read_csv` for this?

Reply:
The reason we did not use cudf's read_csv is because cudf's read_csv adds an 8 byte offset when storing the dataframe into gpu memory. This makes it so that cudf.read_csv can not read a dataframe as large as from_pandas. The solution to this is to read the dataframe one chunk at a time and merge them (dask-cudf does this), but we didn't do that for time reasons. Here's a right up for you to reference.

https://github.com/gatech-sysml/modin/issues/6